File: twitch-frontend.js
// A Twitch frontend for http_backend.py

const crypto = require('crypto');
const url = require("url");
const axios = require("axios");
const tmi = require("tmi.js");

const characterLimit = 500;
const {
    prefix,
    backend_port_number,
    twitch_owner_id,
    twitch_username,
    twitch_token,
    twitch_channels
} = require("./config.json");
const sailorServiceURL = `http://localhost:${backend_port_number}`;
const pastebinURL = "https://ghostbin.com/paste/new";

const clientOptions = {
    identity: {
        username: twitch_username,
        password: twitch_token
    },
    channels: twitch_channels
};

const client = new tmi.client(clientOptions);

client.on("message", onMessage);
client.on("connected", onConnected);

client.connect();

// Truncates an array of strings to a single line for logging
function toOneLiner(data, maxLength=75) {
    let oneLiner = data.join(" ").split("\n").join(" ");
    if (oneLiner.length > maxLength) {
        oneLiner = oneLiner.substring(0, maxLength-1) + "…";
    }
    if (oneLiner.length === 0) {
        oneLiner = "<empty>";
    }
    return oneLiner;
}

/*
Sends a message normally if it's short enough.
Otherwise, posts the message to a pastebin site (currently Ghostbin) and sends the link in chat.
*/
function extSay(channel, tags, data) {
    if (data.length === 0) {
        return;
    }
    else if (data.length === 1) {
        client.say(channel, data[0].split("\n").join(" | "));
    }
    else {
        let pastebinRequestBody = new url.URLSearchParams({
            "text": data.join("\n"),
            "title": "Multiline post",
            "password": crypto
                .randomBytes(32)
                .toString("hex")
        });
        axios
        .post(pastebinURL, pastebinRequestBody.toString())
        .then((response) => {
            let reply = `Multiline post: ${response.request.res.responseUrl}`;
            console.log(`id=${tags.id} | ${reply}`);
            client.say(channel, reply);
        });
    }
}

// Standard ready message
function onConnected(addr, port) {
    console.log(`Connected to ${addr}:${port}`);
}

// Command handler (and potential other stuff)
function onMessage(channel, tags, message, self) {
    if (self) return;
    if (!message.startsWith(prefix)) return;

    console.log(`id=${tags.id} user=${tags.username} userId=${tags["user-id"]} channel=${channel} | ${message}`);

    axios
    .post(sailorServiceURL, {
        "id": `twitch:${tags.id}`,
        "message": message.replace(prefix, "").trim(),
        "is_owner": tags["user-id"] === twitch_owner_id,
        "character_limit": characterLimit
    })
    .then((response) => {
        let replyOneLiner = toOneLiner(response.data);
        console.log(`id=${tags.id} status=${response.status} | ${replyOneLiner}`);
        extSay(channel, tags, response.data);
    })
    .catch((error) => {
        try {
            let replyOneLiner = toOneLiner(error.response.data);
            console.log(`id=${tags.id} status=${error.response.status} | ${replyOneLiner}`);
            extSay(channel, tags, error.response.data);
        }
        catch {
            let errorMessage;
            if (error.code) {
                errorMessage = `An error occurred: ${error.code}`;
            }
            else {
                errorMessage = "An unknown error occurred.";
            }
            console.error(`id=${tags.id} | ${errorMessage}`);
            if (error.code === "ECONNREFUSED") {
                client.say(channel, "My brain stopped working. Please contact my owner. :<");
            }
            else if (error.code !== "ECONNRESET") {
                client.say(channel, errorMessage);
            }
        }
    });
}

File: twitch_frontend.py
"""
A Twitch frontend for http_backend.py

Requires Python 3.6 or higher.
"""

# pylint: disable=invalid-name,broad-except

import asyncio
import json
import secrets

import aiohttp
import twitchio

from sailor_fox.helpers import create_logger, get_prefix, to_one_liner

logger = create_logger(__name__)


def main():
    """Factory to create and start the bot."""

    with open("config.json") as config_file:
        config = json.load(config_file)

    token = config.get("twitch_token")
    assert isinstance(token, str), "OAuth token not valid."

    backend_port_number = config.get("backend_port_number", 9980)
    prefix = str(config.get("prefix", ""))

    client = twitchio.Client(token, initial_channels=config.get("twitch_channels", []))
    client_session = aiohttp.ClientSession(loop=client.loop)

    @client.event()
    async def event_ready():
        """Bot has logged in."""
        logger.info("Logged in as %s", client.nick)

    @client.event()
    async def event_message(message: twitchio.Message):
        """Handle on_message events from Discord and forward them to the processor."""

        prefix_or_none, message_text = get_prefix(message.content, prefix)
        if prefix_or_none is None or not message_text:
            return

        tags = message.tags or {}

        message_id = tags.get("id", secrets.token_hex(16))
        author_id = tags.get("user-id", "unknown")

        is_owner = (author_id == config.get("twitch_owner_id"))

        logger.info(
            "id=%s user=%s userId=%s channel=%s | %s",
            message_id,
            tags.get("display-name", "unknown"),
            author_id,
            message.channel.name,
            to_one_liner(message.content)
        )

        request_body = {
            "id": f"twitch.py:{message_id}",
            "message": message_text,
            "is_owner": is_owner,
            "character_limit": 500,
            "replace_newlines": True
        }

        try:
            async with client_session.post(
                f"http://localhost:{backend_port_number}",
                json=request_body,
                timeout=10
            ) as response:
                action_stack = await response.json()
            is_error = response.status != 200
            reply_stack = []
            for action in action_stack:
                if not isinstance(action.get("type"), str) \
                or not isinstance(action.get("value"), str):
                    continue
                log_message_base = f"id={message_id} actionType={action['type']}"
                if action["type"] == "reply":  # only support reply for now
                    log_message = (
                        f"{log_message_base} | {to_one_liner(action['value'])}"
                    )
                    if is_error:
                        logger.warning(log_message)
                    else:
                        logger.info(log_message)
                    reply_stack.append(action["value"])
                else:
                    logger.warning("%s | Unsupported action", log_message_base)
            for reply_count, reply in enumerate(reply_stack):
                # Hardcode ratelimit for now
                if reply_count > 0:
                    await asyncio.sleep(2)
                await message.channel.send(reply)

        except Exception as error:
            logger.error("id=%s | %s", message_id, to_one_liner(str(error)))
            if isinstance(error, aiohttp.client_exceptions.ClientConnectorError):
                await message.channel.send("My brain stopped working. Please contact my owner. :<")
            else:
                await message.channel.send(str(error))

    client.run()


if __name__ == "__main__":
    main()
